lib/AuthStateManager.ts

```diff
@@ -0,0 +1,148 @@
+  idToken: null,
+  accessToken: null,
+};
+const DEFAULT_PENDING = {
+  updateAuthStatePromise: null,
+  canceledTimes: 0
+};
+const EVENT_AUTH_STATE_CHANGE = 'authStateChange';
+const MAX_PROMISE_CANCEL_TIMES = 10;
+
+class AuthStateManager {
+  _sdk: OktaAuth;
+  _pending: { 
+    updateAuthStatePromise: typeof PCancelable;
+    canceledTimes: number; 
+  };
+  _authState: AuthState;
+
+  constructor(sdk: OktaAuth) {
+    if (!sdk.emitter) {
+      throw new AuthSdkError('Emitter should be initialized before AuthStateManager');
+    }
+
+    this._sdk = sdk;
+    this._pending = { ...DEFAULT_PENDING };
+    this._authState = { ...DEFAULT_AUTH_STATE };
+
+    // Listen on tokenManager events to start updateState process
+    // "added" event is emitted in both add and renew process, it will be canceled by the followed "renewed" event
+    sdk.tokenManager.on('added', (key, token) => {
+      this.updateAuthState({ event: 'added', key, token });
+    });
+    sdk.tokenManager.on('renewed', (key, token) => {
+      this.updateAuthState({ event: 'renewed', key, token });
+    });
+    sdk.tokenManager.on('removed', (key, token) => {
+      this.updateAuthState({ event: 'removed', key, token });
+    });
+  }
+
+  getAuthState(): AuthState {
+    return this._authState;
+  }
+
+  updateAuthState({ event, key, token }: UpdateAuthStateOptions = {}): void {
+    const { isAuthenticated, devMode } = this._sdk.options;
+
+    const logger = (status) => {
+      console.group(`OKTA-AUTH-JS:updateAuthState: Event:${event} Status:${status}`);
+      console.log(key, token);
+      console.log('Current authState', this._authState);
+      console.groupEnd();
+    };
+
+    const emitAuthStateChange = (authState) => {
+      this._authState = authState;
+      // emit new authState object
+      this._sdk.emitter.emit(EVENT_AUTH_STATE_CHANGE, { ...authState });
+      devMode && logger('emitted');
+    };
+
+    if (this._pending.updateAuthStatePromise) {
+      if (this._pending.canceledTimes >= MAX_PROMISE_CANCEL_TIMES) {
+        // stop canceling then starting a new promise
+        // let existing promise finish to prevent running into loops
+        return;
+      } else {
+        this._pending.updateAuthStatePromise.cancel();
+      }
+    }
+
+    /* eslint-disable complexity */
+    const cancelablePromise = new PCancelable((resolve, _, onCancel) => {
+      onCancel.shouldReject = false;
+      onCancel(() => {
+        this._pending.updateAuthStatePromise = null;
+        this._pending.canceledTimes = this._pending.canceledTimes + 1;
+        devMode && logger('canceled');
+      });
+
+      const emitAndResolve = (authState) => {
+        if (cancelablePromise.isCanceled) {
+          resolve();
+          return;
+        }
+        // emit event and clear states
+        emitAuthStateChange(authState); 
+        this._pending = { ...DEFAULT_PENDING };
+        resolve();
+      };
+
+      return this._sdk.tokenManager._getTokens()
```

Comment:
This method sometimes returns a promise and sometimes doesn't. I think it will probably work better if it never returns a promise but is "fire and forget". What would the caller do with a reject?

Reply:
After a closer look, it's actually not been returned, since it's wrapped by the cancelablePromise. But it would be good to remove the `return` to reduce confusion.